bve/src/parse/panel2_cfg/sections/subject.rs

```diff
@@ -0,0 +1,214 @@
+use crate::parse::kvp::FromKVPValue;
+use nom::branch::alt;
+use nom::bytes::complete::{tag, take_while};
+use nom::combinator::{map_res, opt};
+use nom::sequence::preceded;
+use nom::IResult;
+use std::str::FromStr;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Subject {
+    pub target: SubjectTarget,
+    pub digit: Option<u64>,
+}
+
+impl Subject {
+    #[must_use]
+    pub const fn from_target(target: SubjectTarget) -> Self {
+        Self { target, digit: None }
+    }
+}
+
+impl FromKVPValue for Subject {
+    fn from_kvp_value(value: &str) -> Option<Self> {
+        match parse_subject(value) {
+            Ok(("", o)) => Some(o),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum SubjectTarget {
+    Acceleration,
+    Ats(u64),
+    LocoBrakeCylinder,
+    BrakeCylinder,
+    LocoBrakePipe,
+    BrakePipe,
+    Brake,
+    LocoBrake,
+    ConstSpeedSystem,
+    Door,
+    DoorLeft(u64),
+    DoorRight(u64),
+    DoorButtonLeft,
+    DoorButtonRight,
+    EqualizingReservoir,
+    Hour,
+    KilometersPerHour,
+    Minute,
+    MotorAcceleration,
+    MilesPerHour,
+    MainReservoir,
+    MetersPerSecond,
+    PowerNotch,
+    Reverser,
+    StraightAirPipe,
+    Second,
+    True,
+    Klaxon,
+    PrimaryKlaxon,
+    SecondaryKlaxon,
+    MusicKlaxon,
+    PassAlarm,
+    PilotLamp,
+    StationAdjustAlarm,
+}
+
+macro_rules! variant {
+    ($life:lifetime, $name:literal, $variant:ident) => {
+        move |s: &$life str| -> IResult<&$life str, SubjectTarget> {
+            tag($name)(s).map(|(i, _o)| (i, SubjectTarget::$variant))
+        }
+    };
+}
+
+macro_rules! variant_value {
+    ($life:lifetime, $name:literal, $variant:ident, $value:expr) => {
+        move |s: &$life str| -> IResult<&$life str, SubjectTarget> {
+            tag($name)(s).map(|(i, _o)| (i, SubjectTarget::$variant($value)))
+        }
+    };
+}
+
+macro_rules! variant_index {
+    ($life:lifetime, $name:literal, $variant:tt) => {
+        move |s: &$life str| -> IResult<&$life str, SubjectTarget> {
+            preceded(tag($name), parse_number)(s).map(|(i, num)| (i, SubjectTarget::$variant(num)))
+        }
+    };
+}
+
+// noinspection RsNeedlessLifetimes
+fn parse_subject<'a>(input: &'a str) -> IResult<&'a str, Subject> {
+    let (input, target) = alt((
+        alt((
+            variant!('a, "acc", Acceleration),
+            variant_value!('a, "atc", Ats, 271),
+            variant_index!('a, "ats", Ats),
+            variant!('a, "locobrakecylinder", LocoBrakeCylinder),
+            variant!('a, "bc", BrakeCylinder),
+            variant!('a, "locobrakepipe", LocoBrakePipe),
+            variant!('a, "bp", BrakePipe),
+            variant!('a, "brake", Brake),
+            variant!('a, "locobrake", LocoBrake),
+            variant!('a, "csc", ConstSpeedSystem),
+            // These 4 must come before door, as door is a prefix for them and will partially parse, then fail
```

Comment:
10/10 file format vv useful

Reply:
The file format is fine, just how recursive decent parsers work

Reply:
Nope I blame the file format and you can't stop me